### DISCUSSION TITLE
Parallel of pymilvus/Milvus Lite

### DISCUSSION THREAD

**Discussion by BinDong314** [milvus-io/milvus#28918] [Q&A and General discussion]
Hi Milvus,
I am trying to find some information about the parallelization of pymilvus/Milvus Lite . 
Basically, how to control the number of threads on a single node (and also possible way to scale it onto multiple nodes). 

Some information online only show Milvus has some parameters, like max-cpu-available, for the threads numbers.
But, I don't how to access/control the these parameters for a single node.

As said, there is also no doc about about how to scale it on multiple nodes. 
I noticed there is some doc about general  Milvus for multiple nodes here https://milvus.io/docs/install_cluster-milvusoperator.md
Unfortunately, it is not possible for our HPC system where users are not allowed to use K8s 

Thanks.
Bin

I just used milvus in notebook, as below

from milvus import default_server

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility
)
....

**Comment by yhmo** [comment]
milvus_lite is a standalone instance on local, mainly for local testing and prototypes.

For a standalone instance, all the components(proxy, rootcoord, querycoord, datacoord, querynode, datanode, indexnode) are in one process. The internal parallelization is not visible for users, **it is hard to control the number of threads in the standalone.**
Basically, the major parallel tasks:
- for ingestion(insert/delete/upsert), proxy verifies the input and passes the requests to pulsar/kafka, datanode and querynode pull requests from pulsar/kafka and construct segment and data files, rootcoord and datacoord maintain data consistency and "data views".
- for search/query, querycoord schedules the tasks, querynode computes and returns results.
- for index, indexcoord schedules the tasks, indexnode builds index for segments.

**Reply by BinDong314** [reply]
Thanks @yhmo for the response.

What will be default behaviors for milvus_lite without user control on a single node with multiple CPU cores? 
Will it use the all available CPUs/threads for these tasks, such as ingestion/ search/query/index?

**Reply by yhmo** [reply]
The parallel tasks will compete for CPU resources. If an index task is running, and a search request comes in, the search performance will be worse.
